### ISSUE TITLE
User procedure call fails to delete node: EntityNotFoundException/QueryExecutionException

### ISSUE THREAD

**Issue by LeBlue** [neo4j/neo4j#9055]
**Version:** community-3.1.1 
**Operating System:** arch-linux  
**API:** user procedures/cypher

The following example procedure tries to delete a node (`endNode`) related to the input node. The procedure call returns, but the transaction is rolled back with the error `EntityNotFoundException` (HTTP-API) or `QueryExecutionException` (neo4j-shell).

The error occurs depending on how the input nodes are matched in the cypher query.

### Example procedure
```
package example;
import org.neo4j.graphdb.*;
import org.neo4j.procedure.*;

public class example {

    @Procedure(mode = Mode.WRITE)
    @Description("example.testdelete(node, relType) deletes node and related (relType) nodes")
    public void testdelete(@Name("node") Node node, @Name("relType") String relType) {

        Iterable<Relationship> rels = node.getRelationships(RelationshipType.withName(relType), Direction.OUTGOING);
        for (Relationship rel: rels) {
            Node endNode = rel.getEndNode();
            rel.delete();
            endNode.delete(); // this succeeds but transaction with procedure call fails
        }
        node.delete();
    }
}
```

### Steps to reproduce
create a node with related node that should be deleted via the procedure call.
```
create (n1:Label {name: "test"})-[:REL]->(n2) return n1,n2;
+--------------------------------------+
| n1                     | n2          |
+--------------------------------------+
| Node[1]{name:"test"}   | Node[2]{}   |
+--------------------------------------+
```

### Fails
The procedure call fails when accessing node attributes while selecting the nodes to input into the call.
```
match (n) where n.name = "test" call example.testdelete(n, "REL") return null;
QueryExecutionException: Node with id 2 has been deleted in this transaction <- neo4j-shell
{ "code": "Neo.ClientError.Statement.EntityNotFound",
  "message": "Node with id 2 has been deleted in this transaction"  } <- HTTP-API
```
Splitting match and call like this also fails.
```
match (n) where n.name = "test" with n as node call example.testdelete(node, "REL") return null;
````

### Works
With input nodes matched only via patterns that exclude the to be deleted node
```
match (n) where id(n)=1 call example.testdelete(n, "REL") return null;
match (n)-[:REL]->() call example.testdelete(n, "REL") return null;
match (n:Label) where n.name = "test" call example.testdelete(n, "REL") return null;
```

The error seems to be related to the cypher execution order in matching attributes/filtering of a node list and the procedure call.

### Workaround
Collecting all candiate nodes prior to calling `testdelete` works.
```
match (n) where n.name = "test" with collect(n) as ln unwind ln as l call example.testdelete(l, "REL") return null;
```

### Question: bug or limitation

Should this be considered a bug or a limitiation in mixing/interaction of cypher with procedure calls?
Perhaps there is a better/correct way to do this?

Thanks

**Comment by sherfert** [comment]
After over half a year of thoughtful consideration, we came to the conclusion this is a bug. I was able to reproduce it in 3.1.8-SNAPSHOT too. Thanks for the report!

**Comment by craigtaverner** [comment]
We have fixed this in 3.2 and above with a new `eager` attribute for procedures. If the procedure author knows that the procedure is making changes to the database that could influence the behaviour of the incoming data stream, then they should annotate the procedure with `@Procedure( mode = WRITE, eager = true )` and the Cypher planner will plan an eager operation before calling the procedure. It is still necessary for the procedure internal code to deal with any exceptions as a result of self-interaction, but this fix should at least prevent exceptions outside the procedure, like the one reported here which was due to a property lookup in the `n.name = 'test'` predicate.

See https://github.com/neo4j/neo4j/commit/a154bf663565bf7b38e71eb2a8db34838ac9df51
